src/EditorFeatures/Core/FindUsages/ClassifiedSpansAndHighlightSpan.cs

```diff
@@ -0,0 +1,271 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Editor.FindUsages
+{
+    internal struct ClassifiedSpansAndHighlightSpan
+    {
+        private const string Key = nameof(ClassifiedSpansAndHighlightSpan);
+
+        public readonly ImmutableArray<ClassifiedSpan> ClassifiedSpans;
+        public readonly TextSpan HighlightSpan;
+
+        public ClassifiedSpansAndHighlightSpan(
+            ImmutableArray<ClassifiedSpan> classifiedSpans,
+            TextSpan highlightSpan)
+        {
+            ClassifiedSpans = classifiedSpans;
+            HighlightSpan = highlightSpan;
+        }
+
+        public static async Task<DocumentSpan> GetClassifiedDocumentSpanAsync(
+            Document document, TextSpan sourceSpan, CancellationToken cancellationToken)
+        {
+            var classifiedSpans = await ClassifyAsync(
+                document, sourceSpan, cancellationToken).ConfigureAwait(false);
+
+            var properties = ImmutableDictionary<string, object>.Empty.Add(Key, classifiedSpans);
+
+            return new DocumentSpan(document, sourceSpan, properties);
+        }
+
+        public static async Task<ClassifiedSpansAndHighlightSpan> ClassifyAsync(
+            DocumentSpan documentSpan, CancellationToken cancellationToken)
+        {
+            // If the document span is providing us with the classified spans up front, then we
+            // can just use that.  Otherwise, go back and actually classify the text for the line
+            // the document span is on.
+            if (documentSpan.Properties.TryGetValue(Key, out var value))
+            {
+                return (ClassifiedSpansAndHighlightSpan)value;
+            }
+
+            return await ClassifyAsync(
+                documentSpan.Document, documentSpan.SourceSpan, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<ClassifiedSpansAndHighlightSpan> ClassifyAsync(
```

Comment:
all the remaining logic in this type moved from AbstractTableDataSourceFindUsagesContext.cs.  It is unchanged.